From: Peter Lemenkov <lemenkov@gmail.com>
Date: Tue, 24 Jun 2025 11:00:25 +0200
Subject: [PATCH] Don't do vendoring

Signed-off-by: Peter Lemenkov <lemenkov@gmail.com>

diff --git a/Makefile b/Makefile
index b17c76ea70..2b67ff3caa 100644
--- a/Makefile
+++ b/Makefile
@@ -164,7 +164,6 @@ all: erigon $(COMMANDS)
 db-tools:
 	@echo "Building db-tools"
 
-	go mod vendor
 	cd vendor/github.com/erigontech/mdbx-go && MDBX_BUILD_TIMESTAMP=unknown make tools
 	mkdir -p $(GOBIN)
 	cd vendor/github.com/erigontech/mdbx-go/libmdbx && cp mdbx_chk $(GOBIN) && cp mdbx_copy $(GOBIN) && cp mdbx_dump $(GOBIN) && cp mdbx_drop $(GOBIN) && cp mdbx_load $(GOBIN) && cp mdbx_stat $(GOBIN)
