From: Peter Lemenkov <lemenkov@gmail.com>
Date: Sat, 30 Aug 2025 14:19:29 +0200
Subject: [PATCH] Revert "Don't do vendoring"

This reverts commit 526f86a792547f18411152d704e8e7cc11b0b06e.

diff --git a/Makefile b/Makefile
index 2b67ff3caa..b17c76ea70 100644
--- a/Makefile
+++ b/Makefile
@@ -164,6 +164,7 @@ all: erigon $(COMMANDS)
 db-tools:
 	@echo "Building db-tools"
 
+	go mod vendor
 	cd vendor/github.com/erigontech/mdbx-go && MDBX_BUILD_TIMESTAMP=unknown make tools
 	mkdir -p $(GOBIN)
 	cd vendor/github.com/erigontech/mdbx-go/libmdbx && cp mdbx_chk $(GOBIN) && cp mdbx_copy $(GOBIN) && cp mdbx_dump $(GOBIN) && cp mdbx_drop $(GOBIN) && cp mdbx_load $(GOBIN) && cp mdbx_stat $(GOBIN)
