From: Peter Lemenkov <lemenkov@gmail.com>
Date: Mon, 11 Mar 2024 23:08:11 +0100
Subject: [PATCH] Disable silkworm entirely

Signed-off-by: Peter Lemenkov <lemenkov@gmail.com>

diff --git a/Makefile b/Makefile
index 8cc5462ad4..e4e5224b99 100644
--- a/Makefile
+++ b/Makefile
@@ -41,7 +41,7 @@ ifeq ($(shell uname -m),x86_64)
 endif
 
 # about netgo see: https://github.com/golang/go/issues/30310#issuecomment-471669125 and https://github.com/golang/go/issues/57757
-BUILD_TAGS = nosqlite,noboltdb
+BUILD_TAGS = nosqlite,noboltdb,nosilkworm
 
 ifneq ($(shell "$(CURDIR)/turbo/silkworm/silkworm_compat_check.sh"),)
 	BUILD_TAGS := $(BUILD_TAGS),nosilkworm
