From: Alex Sharov <AskAlexSharov@gmail.com>
Date: Wed, 13 Sep 2023 18:15:36 +0700
Subject: [PATCH] =?UTF-8?q?Configurable=20hashers=20amount=C2=A0per=20torr?=
 =?UTF-8?q?ent=20(#867)?=
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit


diff --git a/.gitignore b/.gitignore
index d92c6d5b..0c15585d 100644
--- a/.gitignore
+++ b/.gitignore
@@ -1,3 +1,5 @@
 .idea
 *-run.gob
 .envrc*
+.DS_Store
+go.work*
\ No newline at end of file
diff --git a/config.go b/config.go
index cfcfe392..d779cf90 100644
--- a/config.go
+++ b/config.go
@@ -186,6 +186,8 @@ type ClientConfig struct {
 	ICEServers []string
 
 	DialRateLimiter *rate.Limiter
+
+	PieceHashersPerTorrent int // default: 2
 }
 
 func (cfg *ClientConfig) SetListenAddr(addr string) *ClientConfig {
@@ -223,13 +225,14 @@ func NewDefaultClientConfig() *ClientConfig {
 			Preferred:        true,
 			RequirePreferred: false,
 		},
-		CryptoSelector:        mse.DefaultCryptoSelector,
-		CryptoProvides:        mse.AllSupportedCrypto,
-		ListenPort:            42069,
-		Extensions:            defaultPeerExtensionBytes(),
-		AcceptPeerConnections: true,
-		MaxUnverifiedBytes:    64 << 20,
-		DialRateLimiter:       rate.NewLimiter(10, 10),
+		CryptoSelector:         mse.DefaultCryptoSelector,
+		CryptoProvides:         mse.AllSupportedCrypto,
+		ListenPort:             42069,
+		Extensions:             defaultPeerExtensionBytes(),
+		AcceptPeerConnections:  true,
+		MaxUnverifiedBytes:     64 << 20,
+		DialRateLimiter:        rate.NewLimiter(10, 10),
+		PieceHashersPerTorrent: 2,
 	}
 	cc.DhtStartingNodes = func(network string) dht.StartingNodesGetter {
 		return func() ([]dht.Addr, error) { return dht.GlobalBootstrapAddrs(network) }
diff --git a/torrent.go b/torrent.go
index 5f6ddf39..8b474521 100644
--- a/torrent.go
+++ b/torrent.go
@@ -2251,7 +2251,7 @@ func (t *Torrent) onIncompletePiece(piece pieceIndex) {
 }
 
 func (t *Torrent) tryCreateMorePieceHashers() {
-	for !t.closed.IsSet() && t.activePieceHashes < 2 && t.tryCreatePieceHasher() {
+	for !t.closed.IsSet() && t.activePieceHashes < t.cl.config.PieceHashersPerTorrent && t.tryCreatePieceHasher() {
 	}
 }
 
